### PR TITLE
refactor-ia-microservices

### PR DESCRIPTION
**Pull request by borislav-puzic**
I want to change arhitecture to microservices.
Add ActionBenefitMessage, GrpcAdress, DateRange value objects
Add Tender, PharmacySystem entities

### CODE REVIEW

**Review by milana-todorovic — COMMENTED** [review]
Could you give more meaningful names to the services? Also, I'm not sure why none of the checks seem to be executing on this branch, it's very odd. Maybe because of the conflicts? I don't remember that happening before though.

File: HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/ActionBenefit/ActionBenefitService.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+namespace IntegrationAdapters.MicroserviceComunicator
+{
+    public class ActionBenefitService : IActionBenefitService
+    {
+        private HttpClient _httpClient;
+
+        public ActionBenefitService(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+            _httpClient.BaseAddress = new System.Uri("http://localhost:5002");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/ActionBenefit/ActionBenefitService.cs:19]: This needs to be configurable, either check how we pulled out all the Service urls into config in PatientWebApp or let me know if we need to handle the configuration.

File: HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/Drug/DrugService.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+namespace IntegrationAdapters.MicroserviceComunicator
+{
+    public class DrugService : IDrugService
+    {
+        private HttpClient _httpClient;
+
+        public DrugService(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+            _httpClient.BaseAddress = new System.Uri("http://localhost:5003");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/Drug/DrugService.cs:21]: This also has to be configurable.

File: HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/PharmacySystem/PharmacySystemService.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+namespace IntegrationAdapters.MicroserviceComunicator
+{
+    public class PharmacySystemService : IPharmacySystemService
+    {
+        private HttpClient _httpClient;
+
+        public PharmacySystemService(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+            _httpClient.BaseAddress = new System.Uri("http://localhost:5001");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/PharmacySystem/PharmacySystemService.cs:19]: Configure this.

File: HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/Prescription/PrescriptionService.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+namespace IntegrationAdapters.MicroserviceComunicator
+{
+    public class PrescriptionService : IPrescriptionService
+    {
+        private HttpClient _httpClient;
+
+        public PrescriptionService(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+            _httpClient.BaseAddress = new System.Uri("http://localhost:5001");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/Prescription/PrescriptionService.cs:18]: Configure this.

File: HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/Tender/TenderService.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+namespace IntegrationAdapters.MicroserviceComunicator
+{
+    public class TenderService : ITenderService
+    {
+        private HttpClient _httpClient;
+
+        public TenderService(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+            _httpClient.BaseAddress = new System.Uri("http://localhost:5004");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdapters/MicroserviceComunicator/Tender/TenderService.cs:20]: Configure this.

File: HealthcareSystem/IntegrationAdapters/Startup.cs [hunk 6]
```diff
@@ -85,61 +70,53 @@ public void ConfigureServices(IServiceCollection services)
...
-        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, MyDbContext context, IAntiforgery antiforgery)
+        public void Configure(IApplicationBuilder app, IWebHostEnvironment env, MyDbContext context, IAntiforgery antiforgery, IHttpClientFactory httpClientFactory)
         {
-            app.ApplicationServices.GetService<IRabbitMqTenderingService>();
-            app.ApplicationServices.GetService<IRabbitMqActionBenefitService>();
+            var client = httpClientFactory.CreateClient();
+            client.GetAsync("http://localhost:5001/ping");
+            client.GetAsync("http://localhost:5002/ping");
+            client.GetAsync("http://localhost:5003/ping");
+            client.GetAsync("http://localhost:5004/ping");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdapters/Startup.cs:121]: Configure this.

File: HealthcareSystem/IntegrationAdaptersService1/MicroserviceComunicator/ActionBenefitService.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+namespace IntegrationAdaptersService1.MicroserviceComunicator
+{
+    public class ActionBenefitService : IActionBenefitService
+    {
+        private HttpClient _httpClient;
+        public ActionBenefitService(IHttpClientFactory httpClientFactory)
+        {
+            _httpClient = httpClientFactory.CreateClient();
+            _httpClient.BaseAddress = new System.Uri("http://localhost:5002");
```
Comment (milana-todorovic) [inline HealthcareSystem/IntegrationAdaptersService1/MicroserviceComunicator/ActionBenefitService.cs]: Configure this.

**Review by milana-todorovic — COMMENTED** [review]
I marked out all the urls that were used as constants in the code and need to be made configurable.

**Review by milana-todorovic — COMMENTED** [review]
Also, I think you should move your Service and Repository classes from Backend to the projects that use them (since each should be used by just one service anyway, and we're aiming to make the Backend contain just the data storage model).

**Review by anciz — DISMISSED** [review]

**Review by milana-todorovic — APPROVED** [review]
Seems alright. A separate branch will be made to update the local deployment to fit this architecture, so the configuration will be handled there.

**Review by Stefan-Kitanovic — APPROVED** [review]

**Pull request merged**